Direct, thread-safe, access to thread managed object

ivovandongen opened the pull request:
This pr enables a thread to be suspended and the managed object to be accessed directly. 
- Exposes `thread.hpp`
- Makes thread suspending/resuming thread safe so it can be called from arbitrary threads
- Adds a RAII class that blocks a thread and gives access to the object managed by it (`BlockingThreadGuard`)

Needed for:
- Android a-synchronous rendering: #9576

jfirebaugh commented:
> Makes thread suspending/resuming thread safe so it can be called from arbitrary threads

It doesn't look thread safe to me; there's even an `assert(!paused)` that will fail if a second thread tries to pause an already-paused thread.

ivovandongen commented:
> It doesn't look thread safe to me; there's even an assert(!paused) that will fail if a second thread tries to pause an already-paused thread.

Right. It was definitely missing some pieces. I've fixed up the pause/resume methods and added tests.

jfirebaugh commented:
@ivovandongen This is no longer needed, correct?
